packages/grid/x-data-grid/src/hooks/features/rows/useGridRows.ts

```diff
@@ -325,6 +326,57 @@ export const useGridRows = (
     [apiRef, logger],
   );
 
+  const replaceRows = React.useCallback<GridRowApi['replaceRows']>(
+    (firstRowToRender, lastRowToRender, newRows) => {
+      if (props.signature === GridSignature.DataGrid) {
+        throw new Error(
+          [
+            "MUI: You can't replace rows using `apiRef.current.replaceRows` on the DataGrid.",
+            'You need to upgrade to the DataGridPro component to unlock this feature.',
+          ].join('\n'),
+        );
+      }
+
+      const newRowsIds = newRows.map((row) => row.id);
+      const allRows = gridRowIdsSelector(apiRef);
+      const updatedRows = [...allRows];
+      updatedRows.splice(firstRowToRender, lastRowToRender - firstRowToRender, ...newRowsIds);
+
+      const idRowsLookup = gridRowsLookupSelector(apiRef);
+      const idToIdLookup = gridRowsIdToIdLookupSelector(apiRef);
+      const tree = gridRowTreeSelector(apiRef);
+      const updatedIdRowsLookup = { ...idRowsLookup };
+      const updatedIdToIdLookup = { ...idToIdLookup };
+      const updatedTree = { ...tree };
+      const rowIdsToBeReplaced = allRows.slice(firstRowToRender, lastRowToRender);
+
+      rowIdsToBeReplaced.forEach((id) => {
+        delete updatedIdRowsLookup[id];
+        delete updatedIdToIdLookup[id];
+        delete updatedTree[id];
+      });
+
+      newRows.forEach((row) => {
+        updatedIdRowsLookup[row.id] = row;
+        updatedIdToIdLookup[row.id] = row.id;
+        updatedTree[row.id] = row;
```

Comment:
Tree should contain `GridRowTreeNodeConfig`, not `GridRowModel`.
Unfortunately Typescript doesn't catch that, because we use `any` as default type for `GridRowModel`.

Maybe there's a way to fix this typing issue in v6?
cc @flaviendelangle

Reply:
Nice catch !

We had a discussion about setting `unknown` as the default row model instead of `any` but it was a pain when I tried to do it
We could probably try again at some point

Reply:
I've opened #5734 to address it

Reply:
This is defined correctly. The previous line of `const updatedTree = { ...tree };` forces the correct type.

Reply:
I don't think it's defined correctly. `tree` is of type `GridRowTreeConfig`:

```ts
export type GridRowTreeConfig = Record<GridRowId, GridRowTreeNodeConfig>;

export interface GridRowTreeNodeConfig {
  /**
   * The grid row id.
   */
  id: GridRowId;
  /**
   * The id of the row children.
   * @default []
   */
  children?: GridRowId[];
  /**
   * The id of the footer
   */
  footerId?: GridRowId | null;
  /**
   * The row id of the parent (null if this row is a top level row).
   */
  parent: GridRowId | null;
  /**
   * Current expansion status of the row.
   * @default false
   */
  childrenExpanded?: boolean;
  /**
   * 0-based depth of the row in the tree.
   */
  depth: number;
  /**
   * The key used to group the children of this row.
   */
  groupingKey: GridKeyValue | null;
  /**
   * The field used to group the children of this row.
   * Is `null` if no field has been used to group the children of this row.
   */
  groupingField: string | null;
  /**
   * If `true`, this node has been automatically added to fill a gap in the tree structure.
   * @default false
   */
  isAutoGenerated?: boolean;
  /**
   * Position of the row among its sibling.
   * @default 'body'
   */
  position?: 'body' | 'footer';
  /**
   * If `true`, this row is pinned.
   * @default false
   */
  isPinned?: boolean;
}
```